### PR TITLE
Removed internal Riddler getKey calls

### PR DESCRIPTION
**Pull request by dhr-verma**
## Description

This PR removes internal calls from Historian and Scribe to Riddler to get tenant keys. Instead, it adds an API in Riddler that signs and returns an access token directly. This is needed to support disabling shared access keys for server tenants to support keyless access. With this change access token signing is Riddler's responsibility. It itself managed and signs tokens using private/shared keys instead of exposing the private keys to other services. To maintain backward compatibility, shared keys can still be retrieved using the `getKeys` API.

Overall changes summary:

1) Riddler `/accesstoken` API: This new API will sign an access token for a given tenant. It will decide what keys to use based on the tenant's configuration. If private access keys are enabled, it will prioritize using those keys. Else it will use shared keys.
2) Riddler `/keys` API: This API has been modified to remove the `usePrivateKeys` flag. This flag was intended to be for internal API calls to fetch private keys. However, to support disabling `key` and `secondaryKey` for a tenant, this flag has been removed. Now this API only returns shared keys. If the tenant does not have shared keys, it returns an object where `key1` and `key2` are empty.
3) RIddler `/tenant/:id` API: Adds a new parameter in the create tenant API - `enableSharedKeyAccess`. When this is false, `key` and `secondaryKey` are undefined. Both `enableSharedKeyAccess` and `enablePrivateKeyAccess` cannot be false.
4) Riddler `/keyaccess` API: The `/keylessaccess` API has been modified and made the `/keyaccess` API. This API is used to change a given tenant's key access config. It allows to enable/disable shared key access and private key access. In case shared key access is disabled, `key` and `secondaryKey` are updated to be undefined. Both private and shared key access cannot be false.
5) Historian and Scribe: Removed internal calls to Riddler's `getKeys` API. Instead using the `signToken` API introduced to the `ITenantManager` interface.

## Breaking Changes

Functionally, these changes maintain backward compatibility, i.e. tenants will continue to function the way they have been. However, code wise this introduces some changes that might be breaking:

1) Adds a new API to the `ITenantManager` interface. This is used to make calls to Riddler to return a signed access token.
2) A package bump is needed in Historian base to consume the latest `signToken` API changes. Till then the old `getKeys` API will continue to be used.

## Reviewer Guidance

Changes needed to make `key` and `secondaryKey` optional in the `ITenantDocument` interface. General review of any changes that might break backward compatibility.

### CODE REVIEW

File: server/routerlicious/packages/services/src/deltaManager.ts [hunk 1]
```diff
@@ -60,21 +59,35 @@ export class DeltaManager implements IDeltaService {
...
-	private async getKey(tenantId: string, includeDisabledTenant = false): Promise<string> {
+	private async signToken(
+		tenantId: string,
+		documentId: string,
+		scopes: ScopeType[],
+		includeDisabledTenant = false,
+	): Promise<string> {
 		const tenantManager = new TenantManager(this.authEndpoint, "");
-		const keyP = await tenantManager.getKey(tenantId, includeDisabledTenant);
+		const keyP = await tenantManager.signToken(
```
Comment (znewton) [inline server/routerlicious/packages/services/src/deltaManager.ts:69]: nit: `tokenP`?

File: server/routerlicious/packages/services/src/deltaManager.ts [hunk 2]
```diff
@@ -60,21 +59,35 @@ export class DeltaManager implements IDeltaService {
 		throw new Error("Method not implemented.");
 	}
 
-	private async getKey(tenantId: string, includeDisabledTenant = false): Promise<string> {
+	private async signToken(
```
Comment (znewton) [inline server/routerlicious/packages/services/src/deltaManager.ts:62]: nit: `getSignedToken` or `getAccessToken`?

File: server/routerlicious/packages/services/src/documentManager.ts [hunk 3]
```diff
@@ -122,9 +121,11 @@ export class DocumentManager implements IDocumentManager {
 	}
 
 	private async getBasicRestWrapper(tenantId: string, documentId: string) {
-		const key = await this.tenantManager.getKey(tenantId);
+		const accessToken = await this.tenantManager.signToken(tenantId, documentId, [
+			ScopeType.DocRead,
+		]);
 		const getDefaultHeaders = () => {
-			const jwtToken = generateToken(tenantId, documentId, key, [ScopeType.DocRead]);
+			const jwtToken = accessToken;
```
Comment (znewton) [inline server/routerlicious/packages/services/src/documentManager.ts:128]: nit: this line can just be fully removed I think

**Review by znewton — APPROVED** [review]
Huge fan of this PR. No major changes to request, looks great

File: server/routerlicious/packages/routerlicious-base/src/riddler/api.ts [hunk 4]
```diff
@@ -66,6 +66,30 @@ export function create(
...
+	 * Signs a tenant token using the given tenant's key.
+	 */
+	router.post("/tenants/:id/accesstoken", (request, response) => {
+		const tenantId = request.params.id;
+		const scopes = request.body.scopes;
+		const documentId = request.body.documentId ?? "";
+		const user = request.body.user;
+		const lifetime = request.body.lifetime;
+		const ver = request.body.ver;
+		const jti = request.body.jti;
```
Comment (alexvy86) [inline server/routerlicious/packages/routerlicious-base/src/riddler/api.ts:79]: Stylistic nit, feel free to ignore :). Could also include `documentId` and do the coalescing when passing it to `signToken()` below.
```suggestion
		const { scopes, user, lifetime, ver, jti } = request.body;
		const documentId = request.body.documentId ?? "";
```

**Review by alexvy86 — APPROVED** [review]
Approving for docs. Left a style nit below, but also have a question: the PR description says the `/keys` API will return a 403 if the tenant does not have shared keys; is that accurate? If so, I'd argue it's an incorrect response. _"The HTTP 403 Forbidden error status code means that a server understands a request but refuses to process it"_, most of the time meaning "you are authenticated but not authorized to do this or ask for this". So if a tenant doesn't have shared keys, I think an empty response seems more appropriate? (for a request that is correctly authenticated and authorized).

**Comment by dhr-verma** [comment]
> Approving for docs. Left a style nit below, but also have a question: the PR description says the `/keys` API will return a 403 if the tenant does not have shared keys; is that accurate? If so, I'd argue it's an incorrect response. _"The HTTP 403 Forbidden error status code means that a server understands a request but refuses to process it"_, most of the time meaning "you are authenticated but not authorized to do this or ask for this". So if a tenant doesn't have shared keys, I think an empty response seems more appropriate? (for a request that is correctly authenticated and authorized).

That is a great point tbh! I can modify the return to be empty strings for `key1` and `key2`.

**Pull request merged**